core/src/main/scala/org/apache/spark/eventhubs/PartitionsStatusTracker.scala

```diff
@@ -0,0 +1,379 @@
+  }
+
+  /**
+   * Checks the latest batch with enough updates and retruns the  perforamnce percentage for each partition as a
+   * value between [0-1] where 0 means the partition is not responding and 1 means it's working wihtout any
+   * perforamnce issue. This information can be used to adjust the batch size for each partition in the next batch.
+   */
+  def partitionsPerformancePercentage(): Option[Map[NameAndPartition, Double]] = {
+    // if there is no batch in the tracker, return None
+    if(batchesStatus.isEmpty) {
+      logDebug(s"There is no batch in the tracker, so return None")
+      None
+    }
+    else {
+      // find the latest batch with enough updates
+      // In Scala 2.13 we can use: val latestUpdatedBatch = batchesStatus.maxByOption(b => b._2.receivedEnoughUpdates)
+      implicit val ordering = new Ordering[(Long, BatchStatus)] {
+        override def compare(x: (Long, BatchStatus), y: (Long, BatchStatus)): Int = (x._1 - y._1).toInt
+      }
+      val batchesWithEnoughUpdates = batchesStatus.filter(b => b._2.receivedEnoughUpdates)
+      val latestUpdatedBatch: Option[BatchStatus] =
+        if (batchesWithEnoughUpdates.isEmpty) None else Some(batchesWithEnoughUpdates.max._2)
+
+
+      latestUpdatedBatch match {
+        case None => {
+          logDebug(s"No batch has ${PartitionsStatusTracker.enoughUpdatesCount} partitions with updates (enough updates), " +
+            s"so return None")
+          None
+        }
+        case Some(batch) => {
+          logDebug(s"Batch ${batch.batchId} is the latest batch with enough updates. Caculate and return its perforamnce.")
+          val performancePercentages = batch.getPerformancePercentages
+          PartitionsStatusTracker.throttlingStatusPlugin.foreach(
+            _.onPartitionsPerformanceStatusUpdate ( batch.batchId,
+                                                    batch.paritionsStatus.map(par => (par._1, par._2.batchSize))(breakOut),
+                                                    batch.paritionsStatus.map(par => (par._1, par._2.batchReceiveTimeInMillis))(breakOut),
+                                                    performancePercentages)
+          )
+          performancePercentages
+        }
+      }
+    }
+  }
+
+  /**
+   * Clean up the tracker. This will be called when the source has been stopped
+   */
+  def cleanUp() = {
+    batchesStatus.map(b => b._2.paritionsStatus.clear)
+    batchesStatus.clear
+    partitionSeqNoPairToBatchIdMap.clear
+  }
+
+  /**
+   * This methods i being used for testing
+   */
+  def batchIdsInTracker: scala.collection.Set[Long] = {
+    this.batchesStatus.keySet
+  }
+}
+
+object PartitionsStatusTracker {
+  private val _partitionsStatusTrackerInstance = new PartitionsStatusTracker
+  val BATCH_NOT_FOUND: Long = -1
+  val ACCEPTABLE_RECEIVE_TIME_PER_EVENT_MS: Double = 1.0
+  val ACCEPTABLE_DIFFERENCE_TIME_PER_EVENT_MS: Double = 1.0
+  var acceptableBatchReceiveTimeMS: Long = 500
```

Comment:
Let's set the default to 30 seconds. Please note that default message count will be 1000 per partition.

Reply:
Use DefaultMaxAcceptableBatchReceiveTime

Reply:
Done